### PR TITLE
Add Style Docs

### PR DESCRIPTION
**Pull request by birkskyum**
Closes https://github.com/maplibre/maplibre-gl-js-docs/issues/400

Preview here: https://birkskyum.github.io/maplibre-style/

There are many reasons to abandon the existing docs, but mainly we need proper typescript support. [MapBox Batfish](https://github.com/mapbox/batfish) only supports javascript, and while it forces us to do a ton of workarounds we still end up with bugs like [421](https://github.com/maplibre/maplibre-gl-js-docs/issues/421) due to lacking ts support.

These new docs are built on a clean modern stack, and has no dependency on mapbox tools - batfish/drui/mrui.

The docs can be launched by running:

cd docs
npm i
npm run dev

### CODE REVIEW

**Comment by wipfli** [comment]
Looks really nice, thanks a lot @birkskyum 

image

The URLs will still be the same as before, right?

**Comment by HarelM** [comment]
Can you please share some visuals of the new docs site?
Can you elaborate on the selection of the tools? Pros? cons? Why was this selected?
Can you please share what should be the process when updating the style-spec? Will it be fully automatic or will there be manual changes that need to happen here?

File: docs/README.md
Comment (HarelM) [inline docs/README.md:1]: Is this readme relevant? Looks weird...

**Comment by wipfli** [comment]
image

File: docs/src/components/appropriate-image/appropriate-image.tsx
Comment (HarelM) [inline docs/src/components/appropriate-image/appropriate-image.tsx:1]: Is these comment out code needed?

**Comment by wipfli** [comment]
Clicking on things like "string" does not work yet: http://localhost:3000/maplibre-gl-style-spec/style-spec/types/#string

**Comment by birkskyum** [comment]
The URLs will be exactly the same as before. The only one I changed was /other -> /deprecations because it's more informative of what the page contains.

Here are some visuals of the menus - first on medium screens the menu one on the right will be pushed into a collapsible part on top of the main content, then on small screen the left menu will be hidden as well and toggled from the header.

Screenshot 2023-04-03 at 22 25 22
Screenshot 2023-04-03 at 22 25 11
Screenshot 2023-04-03 at 22 25 35
Screenshot 2023-04-03 at 22 25 42
Screenshot 2023-04-03 at 22 25 53
Screenshot 2023-04-03 at 22 26 04

File: docs/src/components/items/item.tsx
Comment (HarelM) [inline docs/src/components/items/item.tsx:35]: This is exactly what I would like to avoid - the need to change this code when changing the spec file.

**Comment by birkskyum** [comment]
@wipfli  can you elaborate what links we're talking about?

**Comment by wipfli** [comment]
Go to http://localhost:3000//layers#paint-background-background-opacity, click on "number" -> page not found

image

**Comment by wipfli** [comment]
I think basically not all titles in the "types" page have links.

File: docs/src/routes/types.tsx
Comment (HarelM) [inline docs/src/routes/types.tsx:4]: This is missing the point of having markdown files that can be edited in github and previewed. Instead the text is hard to get right.

Comment (birkskyum) [reply docs/src/routes/types.tsx:4]: Before shooting this down entirely, remember that the website has HMR (npm run dev) with a near-instant preview. It's excellent for editing.

**Comment by HarelM** [comment]
There's just too much code here from my point of view for a documentation site.
We might as well create a site in WordPress or Wix.
The main point from my point of view is how I can easily convert the v8.json file into a site.
This doesn't solve this issue for me.
If this is only a temporary solution, to have a doc site here so that we can work on the next gen, then this is great, otherwise it misses the point I think...

File: docs/src/routes/types.tsx
Comment (HarelM) [reply docs/src/routes/types.tsx:4]: But then why use markdown? Simply code the relevant html/react here...

File: docs/README.md
Comment (birkskyum) [reply docs/README.md]: it's not

File: docs/src/components/appropriate-image/appropriate-image.tsx
Comment (birkskyum) [reply docs/src/components/appropriate-image/appropriate-image.tsx]: No

**Comment by birkskyum** [comment]
@HarelM , the docs repo contains logic to parse the reference .json, and that is what makes it hard to reduce the code size further. I've kept the layouts/SSG to a minimum, but the parsing logic, contained within the /expression folder, Item/Items, is gonna be around in Wordpress too to generate the content.

File: docs/src/routes/types.tsx
Comment (birkskyum) [reply docs/src/routes/types.tsx:4]: The stack can handle a mix of both, but I think there in many cases are quite a lot more boilerplate when using html

Comment (birkskyum) [reply docs/src/routes/types.tsx:4]: I initially took the .mdx approach with astro. Markdown with some JSX components. It worked for some things, but then we have all the code highlighting, which in markdown is indicated with a string. The problem is that we can't use template literals in Markdown, so having it this way we can easily inject variables in the markdown, and even the code snipped which MDX/MD doesn't allow for. That means that in fully markdown codebase, we would have to make a lot more (React) components for all the cases where Markdown fall short, and it's not pretty, so that's why I ended up in this approach.

**Comment by birkskyum** [comment]
> We might as well create a site in WordPress or Wix.

This comment is just ridiculous, but I'll wish you happy coding if you down that path.

**Comment by birkskyum** [comment]
The selected stack is:

Vite - it's the best dev tooling out there. Easy bundling, snappy HMR, huge and growing ecosystem. My impression is that it's the most common approach.

SolidStart for SSG - it basically just a plugin on top of Vite, that allow for File routing, so we have the pages as separate files in a folder, and those file names are the urls we'll get - again simple and low maintenance.

Solid for component hierarchies - it could be anything really as there is almost no runtime logic.

SCSS / CSS modules for styling. This could become just css very soon as Chrome 112 has nested css rules, but it's not out yet. My impression is that it's the most common approach.

**Comment by roblabs** [comment]
@birkskyum  — It's really nice.  I welcome the move to the "solid stack" of modern tools.  It builds easy enough for me and as Oliver noted, I too found a few broken links.  I think those will get sorted.   The documentation shortcuts (see below) are really nice, and I believe preserve what we had with batfish.

Overall, I like the documentation upgrade.

**Comment by louwers** [comment]
From the documentation of SolidStart:

> We are very excited about SolidStart, but it is currently in beta. There will very likely be **bugs and missing documentation**. Come at it with a growth attitude, have fun, and don't use it yet for anything critical.

**Comment by birkskyum** [comment]
@louwers , I had all the routes beneath a /en/ url path initially, would that be a suitable internationalization strategy to you going forward? So that we could add /xx/ languages down the line.

**Comment by birkskyum** [comment]
> Tightly coupled to v8 JSON representation which I thought we wanted to move away from.

This can still be the plan.

**Comment by birkskyum** [comment]
> I was unable to build but I think lack of versioning may also a problem.

What error did you get?

**Comment by birkskyum** [comment]
> As a web developer this all looks pretty familiar to me, but having to learn a JS framework and JS build tools makes contributing to the documentation less approachable.

I've yet to see a SSG setup without a learning curve - what alternative did you have in mind for comparison?

**Comment by birkskyum** [comment]
> From the documentation of SolidStart:
> 
> > We are very excited about SolidStart, but it is currently in beta. There will very likely be **bugs and missing documentation**. Come at it with a growth attitude, have fun, and don't use it yet for anything critical.
> 
> Sounds a little too bleeding edge to me...

I understand why that message might alarm you, but we don't use any of the very experimental features (trps, data loading, island routing etc.) for this setup.

**Comment by birkskyum** [comment]
> we don't have the capacity to keep a partly home grown JS+HTML+CSS stack up to date. 

I've spent weeks maintaining the stack of the batfish docs, which causes headaches for every new version of node, typescript etc. that's released. I even had to become a maintainer of https://github.com/documentationjs/documentation to get the job done. Therefore I think it's safe to say that this is a lot cheaper than keeping what we have.

**Comment by birkskyum** [comment]
> I am unable to build:

 @louwers  Did you run `npm i` first?

**Comment by birkskyum** [comment]
Thanks for all the feedback. I will look into fixing the links. The other mentioned features/refactors are good suggestions, but they go beyond the capabilities of the existing site, and they weren't part of the bounty scope, so it'll have to happen in future efforts.

File: docs/src/components/caption.tsx
Comment (HarelM) [inline docs/src/components/caption.tsx:1]: Can we improve the prop type here?

File: docs/src/components/header/header.module.scss
Comment (HarelM) [inline docs/src/components/header/header.module.scss:4]: Consider using kebab casting here

File: docs/src/components/caption.tsx
Comment (birkskyum) [reply docs/src/components/caption.tsx:1]: Yes, all of these 'children' should be set {children?: JSX.Element}

File: docs/src/components/items/item.tsx
Comment (HarelM) [inline docs/src/components/items/item.tsx:3]: Can we remove this comment?

File: docs/src/components/markdown/markdown.tsx
Comment (HarelM) [inline docs/src/components/markdown/markdown.tsx:28]: Remove comments.

**Comment by HarelM** [comment]
I've reviewed most of this PR.
Please do a comments cleanup and see where you can improve the types.
Also please see if you can align the css to a single casing style and fix the code accordingly.
I would consider using a css framework like tailwind or bootstrap (not sure what's the status of bootstrap these days) - feels like there is too much css files here, this can be split into a different issue/PR if needed.
Overall this is a step forward in terms of running this locally and dependency reduction.
This is approved in general.

**Review by HarelM — APPROVED** [review]
Cleanup a bit before merging this.

**Comment by HarelM** [comment]
Screenshot_20230406_123137_Chrome
Wrong indentation, links should not be to mapbox

**Comment by HarelM** [comment]
Screenshot_20230406_123355_Chrome
Weird exclamation marks

**Comment by birkskyum** [comment]
> Weird exclamation marks

Indeed. Will make sure it doesn't appear in the final build - it's hydration markers that cause it.

**Comment by birkskyum** [comment]
> Wrong indentation, links should not be to mapbox

Where can I point the "sprite" and "glyphs"?

**Comment by HarelM** [comment]
Demotiles repo I think. Check out the style there.

**Comment by birkskyum** [comment]
The code should be ready now.

File: docs/src/expression/expression-metadata.tsx
Comment (birkskyum) [reply docs/src/expression/expression-metadata.tsx:7]: There's a PR here so we remember
https://github.com/maplibre/maplibre-style/issues/89

**Pull request merged**
